doc/man/man1/recode.1
\"    This documentation is copyright 1994 David Huron.
.TH recode 1 "1994 Dec. 4"
.AT 3
.sp 2
.SH "NAME"
.in +2
.in +12
.ti -12
\fBrecode\fR  --  recode numeric tokens in selected Humdrum spines
.in -12
.in -2
.sp 1
.sp 1
.SH "SYNOPSIS"
.in +2
.in +7
.ti -7
\fBrecode\fR  -f  \fIreassign-file\fR  \fB-i  '\fI**interpretation\fB'  [\fB-s \fIregexp\fR]  [\fB-x\fR]  [\fIinputfile\fR ...]
.in -7
.in -2
.sp 1
.sp 1
.SH "DESCRIPTION"
.in +2
The
.B "recode"
command is used to recode numeric components of data tokens in
selected input spines.
Typically,
.B "recode"
is used to reassign a range of numerical values into a finite set of classes
or categories.
For example,
.B "recode"
could be used to reassign all numerical values less than zero to the value -1,
and to assign all values greater than or equal to zero to the value +1.
A typical use of
.B "recode"
might be to reassign melodic intervals (represented in semitones) to one of
five categories: (1) unison [0 semits],
(2) up step [plus 1 or 2 semits], (3) up leap [plus 3 or more semits],
(4) down step [minus 1 or 2 semits], (5) down leap [minus 3 or more semits].
Similarly, duration information might be rhythmically \(odjustified\(cd
so that all durations near 0.5 seconds are recoded as precisely 0.5 seconds.
.sp 1
.sp 1
Note that
.B "recode"
will modify only those input spines matching the exclusive interpretation
specified in the command line.
.sp 1
.sp 1
The manner by which numeric values are reassigned is specified by
the user in a separate
.I "reassignment file."
Reassignment files consist of one or more reassignment records;
each record specifies a
.I "condition"
and a resulting replacement
.I "string."
When the condition is satisfied, the numerical data is replaced by
the associated string.
A simple reassignment file is:
.in +2
.sp 1
.TS
l l.
==0	zero
!=0	other
.TE
.sp 1
.in -2
This file contains two reassignment records.
.I "Conditions"
are given in the left column and the associated replacement
.I "strings"
are given in the right column.
Conditions and strings are separated by a single tab.
Given the above reassignment, when a numerical value in an input token
is equal to zero, the output replaces the input number by the
alphabetic string \(od\f(CRzero\fR.\(cd
The second condition (!= means not-equals) indicates that if a
numerical value not equal to zero is encountered in an input token,
the output replaces the number by the alphabetic string
\(od\f(CRother\fR.\(cd
.sp 1
.sp 1
Permissible relational operators are listed in the following table.
.sp 1
.sp 1
.sp 1
.TS
l l.
==	equals
!=	not equals
<	less than
<=	less than or equal
>	greater than
>=	greater than or equal
else	default relation
.TE
.sp 1
.ce
.I "Relational operators for \fBrecode\fI"
Permissible replacement strings include any combination of printable ASCII
characters with the exception of the tab.
.sp 1
.sp 1
Conditions are tested in the order given in the reassignment file.
Thus if a numeric value satisfies more than one condition,
only the first string replacement is made.
Consider, for example, the following reassignment file:
.in +2
.sp 1
.TS
l l.
<=0	LOW
>100	HIGH
>0	MEDIUM
.TE
.sp 1
.in -2
In this case, all numeric values are replaced by one of three strings:
HIGH, MEDIUM,
or
LOW.
The order of specification is important in the above file.
If the \f(CRMEDIUM\fR condition was specified prior to the \f(CRHIGH\fR
condition, then all values greater than one hundred would be categorized
as \f(CRMEDIUM\fR rather than as \f(CRHIGH\fR.
.sp 1
.sp 1
The
.I "else"
relation can be used to specify the default string output for
numeric input values that satisfy none of the preceding conditions
in the reassignment file.
If no
.I "else"
condition is specified and none of the other conditions are satisfied,
.B "recode"
outputs the original input token without any modification.
.sp 1
.sp 1
Substitutions are made even when a number
is embedded in non-numeric data.
For example, given the above reassignment file, an input token
\f(CRfoo200bar\fR would be output as \f(CRfooHIGHbar\fR.
That is, the numeric portion of the input string (200) would be
deemed to satisfy the condition (>100) and so would be replaced by
the string ("HIGH").
.sp 1
.sp 1
An important property of the
.B "recode"
command is that string replacements are limited to the
.B "first"
occurrence of numeric data within each data token.
Subsequent numeric data within the token remains untouched.
Thus, using the above reassignment file, the input token
\f(CRfoo200bar300\fR
would be output as
\f(CRfooHIGHbar300\fR.
.sp 1
.sp 1
In the case of multiple-stops (data tokens having two or more parts
separated by spaces),
.B "recode"
processes the first occurrence of numeric data for each part of the token.
For example, the double-stop token \f(CRfoo200 bar300\fR
would be output as \f(CRfooHIGH barHIGH\fR.
.sp 1
.sp 1
The
.B "recode"
command provides options to identify which data tokens may be excluded
(skipped) in processing
.B "(-s),"
plus an option that suppresses the echoing of unprocessed signifiers
in the output
.B "(-x)."
See OPTIONS for further information.
.in -2
.sp 1
.sp 1
.SH "OPTIONS"
.in +2
The
.B "recode"
command provides the following options:
.sp 1
.TS
l l.
\fB-f \fIreassign\fR	use reassignments given in file \fIreassign\fR
\fB-h\fR	displays a help screen summarizing the
	  command syntax
\fB-i \fI'**interp'\fR	process only \fI**interp\fR spines
\fB-s \fIregexp\fR	skip; completely ignore tokens matching
	  \fIregexp\fR; (echo in output only)
\fB-x\fR	(exclude) do not echo unprocessed data
	  signifiers in the output
.TE
.sp 1
Options are specified in the command line.
.sp 1
.sp 1
The user can suppress the echoing of non-numeric data within a token
by specifying the
.B "-x"
option on the command line.
When this exclude option is selected, only the replacement strings are output.
For example, given the following reassignment file:
.in +2
.sp 1
.TS
l l.
<=0	LOW
>100	HIGH
>0	MEDIUM
.TE
.sp 1
.in -2
The input token
foo200bar
would be output as
\f(CRHIGH\fR.
If a data token contains no numeric component, then the
.B "-x"
option causes a null token to be output.
.sp 1
.sp 1
The
.B "-x"
option also suppresses the echoing of unprocessed numerical components.
(Recall that string replacements made by
.B "recode"
are limited to the first occurrence of numerica data within a data token.)
For example, with the
.B "-x"
option, the input data token
foo200bar17
would be output as
\f(CRHIGH\fR.
.sp 1
.sp 1
Processing of certain types of data tokens may be avoided by invoking the
.B "-s"
(skip) option.
This option must be accompanied by a user-define regular-expression
(see \fBregexp (6)\fR).
Input data tokens matching this expression are not processed
and are simply echoed in the output.
This option may be useful, for example, in avoiding the processing
of barlines, or other types of data.
.in -2
.sp 1
.sp 1
.SH "EXAMPLES"
.in +2
The operation of the
.B "recode"
command can be illustrated by referring to the following hypothetical
Humdrum file named \f(CRpatrie\fR.
.sp 1
.sp 1
.in +2
.sp 1
.TS
l l.
**kern	**abc
16g	0
8.g	00
16g	1
\(eq1	\(eq1
4cc	2.0
4cc	+3.
4ee	4
4ee	-1
\(eq2	\(eq2
4.gg	22.
8ee	1.1
8.cc	.1
16cc	x1X
8.ee	x1x2x
16cc	1 2
\(eq3	\(eq3
4a	.
4r	r
*-	*-
.TE
.sp 1
.in -2
Consider also the following \(odreassignment\(cd file, named \f(CRreassign\fR.
.in +2
.sp 1
.TS
l l.
==0	zero
==1	one
==2	two
<0	negative
<=3	<=3
>4	>4
else	???
.TE
.sp 1
.in -2
The command:
.sp 1
.sp 1
.in +2
recode -s = -i '**abc' -f reassign patrie
.in -2
.sp 1
.sp 1
would produce the following output:
.in +2
.sp 1
.TS
l l.
**kern	**abc
16g	zero
8.g	zero
16g	one
\(eq1	\(eq1
4cc	two
4cc	<=3
4ee	???
4ee	negative
\(eq2	\(eq2
4.gg	>4
8ee	<=3
8.cc	<=3
16cc	xoneX
8.ee	xonex2x
16cc	one two
\(eq3	\(eq3
4a	.
4r	r
*-	*-
.TE
.sp 1
.in -2
Notice the following:
(1) the measure numbers \f(CR1\fR and \f(CR2\fR have remained
unchanged due to the skip option \f(CR-s =\fR,
(2) the input \f(CRx1X\fR has been replaced by the output string
xoneX
(non-numeric data remain in the same relative position),
(3) the input \f(CRx1x2x\fR has been replaced by the output string
xonex2x
(only the first numerical value in each token is modified),
(4) the double-stop input \f(CR1 2\fR has been replaced by the output string
one two,
and (5) both the null token (.) and the non-numeric token (\f(CRr\fR)
have been echoed in the input unchanged.
.sp 1
.sp 1
Note that with the
.B "-x"
option, all of the non-numeric signifiers in
**abc
spine would be suppressed in the output.
The single non-numeric token (\f(CRr\fR) would be output as a null token.
.in -2
.sp 1
.sp 1
.SH "PORTABILITY"
.in +2
\s-1DOS\s+1 2.0 and up, with the \s-1MKS\s+1 Toolkit.
\s-1OS/2\s+1 with the \s-1MKS\s+1 Toolkit.
\s-1UNIX\s+1 systems supporting the
.I "Korn"
shell or
.I "Bourne"
shell command interpreters, and revised
.I "awk"
(1985).
.in -2
.sp 1
.sp 1
.SH "SEE ALSO"
.in +2
\fBhumsed\fR (1), \fBrend\fR (1), \fBregexp\fR (1),
\fBsed\fR (\s-1UNIX\s+1)
.in -2


toolkit/awk/recode.awk
#####################################################################
##					RECODE.AWK
# Programmed by: Tim Racinsky        Date: July, 1993
# Copyright (c) 1993 David Huron
#
# Modifications:
#   Date:      Programmer:       	Description:
# May 9/94	Tim Racinsky		Fixed 'Eating Backslash' problem (recoded
#							tokens needed an extra '\' added)
# June 7/94	Tim Racinsky		Modified parse_command to work with getopts
#
#
#	This program is called from recode.ksh and is used to 're-code' any
# tokens that contain numeric values for a given exclusive interpretation
# based on a file provided by the user.
#
#	The following gives a brief summary of the available options:
#
#	-s reg_exp: allows user to skip over lines with the given regular expres.
#	-x: suppresses output of non-processed data signifiers
#
# Main Functions used:
#	Parse_command()		Store_indicators()		Store_new_interps()
#	Process_indicators()	Ins_array_pos()		Del_array_pos()
#	Exchange_spines()		Process_data()			Recode()
#
#	VARIABLES:
#
#  -stderr : allows errors to be printed to standard error.
#  -number : regular expression for a valid number.
#  -rel_reg_exp : regular expression for a valid relation
#  -options : holds options specified by user.
#  -reg_exp :  holds regular expression specified by user (if any).
#  -target_interp : holds name of exclusive target interpretation
#  -default_output : holds value of default action.
#  -num_of_conditions : holds number of different conditions to test.
#  -current_no_of_spines : holds current number of active spines.
#  -no_interpretation_yet : flag to indicate if first interp. found yet.
#
#	ARRAYS:
#
#  -current_interp and path_indicator are used to store exclusive
#   interpretations and spine path indicators respectively.
#
BEGIN {
	USAGE="\nUSAGE: recode -h                 (Help Screen)\n       recode "\
		 "[-x] [-s reg_exp] -i **interpretation -f reassign [file ...]"
	#
	# Set any necessary flags, variables, and constants here
	#
	FS = OFS = "\t"
	TRUE = 1; FALSE = 0
	#
	# Set standard error to the appropriate 'file'
	#
	#if (ARGV[0] == "nawk") stderr = "/dev/tty"
	#else stderr = "con"
	"echo $CON" | getline stderr
	close("echo $CON")
	number = "[+-]?(([0-9]+(\\.[0-9]*)?)|((0*)?\\.[0-9]+))"
	rel_reg_exp = "^((==)|(>=)|(<=)|(>)|(<)|(!=))"
	#
	options = ""
	reg_exp = ""
	target_interp = ""
	default_output = ""
	num_of_conditions = 0
	#
	# Determine the options specified by the user, passed from the shell script
	#
	parse_command()
	}
{
#
# Reset variables for each new file
#
if (FNR == 1)
	{
	current_no_of_spines = 0
	no_interpretation_yet = TRUE
	}
#
# Print all comments
#
if ($0 ~ /^!/) print $0
#
# If the current record is an interpretation record
#
else if ($0 ~ /^\*/)
	{
	#
	# Always print it
	#
	print $0
	#
	# Process the first interpretation record.
	#
	if (no_interpretation_yet)
		{
		current_no_of_spines = NF
		no_interpretation_yet = FALSE
		store_new_interps()
		}
	#
	# Otherwise, check to see if it is a spine-path record
	#
	else
		{
		spine_path_record = FALSE
		for (i = 1; i <= current_no_of_spines ; i++)
			{			
			if ($i ~ /^\*\+$|^\*\-$|^\*\^$|^\*v$|^\*x$/) {
				spine_path_record = TRUE 
				break
				}
			}
		#
		# If it is a spine-path record, store and process the indicators
		#
		if (spine_path_record)
			{
			store_indicators()
			process_indicators()
			#
			# Reset any necessary variables if all spines terminate
			#
			if (current_no_of_spines == 0)
				{
				no_interpretation_yet = TRUE
				}
			}
		#
		# Otherwise, process the interpretations
		#
		else store_new_interps()
		}
	}
#
# Otherwise, process the data record
#
else process_data()
}

############################################################################
#				FUNCTIONS USED IN RECODE.AWK
############################################################################

############################################################################
##				Function Parse_command
# This function inspects the array ARGV in order to determine the correct
# options that the user has specified.
#
function parse_command()
	{
	#
	# Process any necessary variables
	#
	if (ARGV[2] != "null") options = ARGV[2]
	if (ARGV[3] != "null") reg_exp = ARGV[3]
	if (ARGV[4] !~ /^\*\*/)
		{
		print "recode: ERROR: Target interpretation must be"\
			 " exclusive" > stderr
		exit
		}
	target_interp = ARGV[4]
	#
	# Read in the "re-assignment" file and parse into three
	# arrays: relation, condition, and action.
	# (Awk automatically generates an error
	# message if the input file is not found.)
	#
	while (getline < ARGV[5] > 0)
		{
		#
		# Parse each input line into three elements:
		# 	relation[]   - i.e. <  <=  >  >=  !=  ==
		#	condition[]  - numeric test, e.g. "3" in "==3"
		#	action[]     - corresponding output when test is true
		#
		# All 3 elements must be present or an error is issued
		#
		if ($1 ~ /^(else)$/) default_output = $2
		else
			{
			if (match($1,rel_reg_exp))	
				{
				relation[++num_of_conditions] = \
									substr($1,RSTART,RLENGTH)
				if (match($1,number))
					{
					condition[num_of_conditions] = \
									substr($1,RSTART,RLENGTH)+0
					if ($2 != "")
						{
						action[num_of_conditions] = $2
						}
					else
						{
						print "recode: ERROR: No action specified " \
							 "on line " num_of_conditions " in " \
							 "reassign file" > stderr
						exit
						}
					}
				else
					{
					print "recode: ERROR: No condition specified " \
						 "on line " num_of_conditions " in " \
						 "reassign file" > stderr
					exit
					}
				}
			else
				{
				print "recode: ERROR: No relation specified " \
					 "on line " num_of_conditions+1 " in " \
					 "reassign file" > stderr
				exit
				}
			}
		}
	ARGV[1] = ARGV[2] = ARGV[3] = ARGV[4] = ARGV[5] = ""
	}

################################################################
##			Function Store_new_interps
#	This function stores the new interpretations found in an
# interpretation record for each spine.
#
function store_new_interps(      j)
	{
	#
	# Loop through each valid spine and store exclusive interpretations
	#	
	for (j = 1; j <= current_no_of_spines; j++)
		{
		if ($j ~ /^\*\*/) current_interp[j] = $j
		}
	}

##############################################################
##			Function Store_indicators
#	This function allows the spine-path indicators for the 
# current record to be stored in the array 'path_indicator' so
# that they may be used later.
#
function store_indicators(   i)
	{
	for (i = 1; i <= current_no_of_spines; i++)
		{
		#
		# If the current field is a null interpretation
		#
		if ($i == "*") path_indicator[i] = "*"
		#
		# If the current field is a spine-path indicator, store it
		#
		else path_indicator[i] = substr($i,2,1)
		}
	} 

##################################################################
##				Function Process_indicators
#	This function takes the spine-path indicators that were stored
# in the array 'path_indicator' in the function 'store_indicators'
# and manipulates the arrays 'path_indicator' and 'current_interp'
# according to the contents of the array 'path_indicator'.
#
function process_indicators(    i)
	{
	i = 1
	#
	# Loop through the entire array of 'path_indicator'
	# NOTE: This array may be altered during the process
	#
	while (i <= current_no_of_spines)
		{
		#
		# If a spine-split indicator occurred
		#
		if (path_indicator[i] == "^")
			{
			#
			# Insert a new array position in 'path_indicator'
			# and 'current_interp'
			#
			ins_array_pos(i+1)
			path_indicator[i+1] = "*"
			current_interp[i+1] = current_interp[i]
			}
		#
		# If a spine-termination indicator occurred
		#
		if (path_indicator[i] == "-")
			{
			#
			# Delete that spine from the arrays and look at the same
			# position the next time through the loop
			#
			del_array_pos(i)
			i--
			}
		#
		# If a spine-exchange indicator occurred
		#
		if (path_indicator[i] == "x")
			{
			#
			# Exchange the paths
			#
			exchange_spines()
			}
		#
		# If a join-path indicator occurred
		#
		if (path_indicator[i] == "v")
			{
			del_array_pos(i+1)
			#
			# If the next indicator is not a join-path indicator
			# then return to original conditions
			#
			if (path_indicator[i+1] != "v")
				{
				path_indicator[i] = "*"
				}
			#
			# Otherwise, look at current and next paths to join
			#
			else i--
			}
		#
		# If a new-path indicator occurred
		#
		if (path_indicator[i] == "+")
			{
			ins_array_pos(i+1)
			current_interp[i+1] = ""
			}
		i++
		}
	}

############################################################
##			Function Ins_array_pos
#	This function inserts a new position in the arrays 'path_indicator'
# and 'current_interp' and copies elements so that order is preserved
#
function ins_array_pos(current_element,     j) 
	{
	for (j = current_no_of_spines+1; j > current_element; j--)
		{
		path_indicator[j] = path_indicator[j-1]
		current_interp[j] = current_interp[j-1]	
		}
	path_indicator[current_element-1] = "*"
	path_indicator[current_element] = "*"
	current_no_of_spines++
	}

##############################################################
##			Function Del_array_pos
#	Performs the opposite of function 'ins_array_pos'.
#
function del_array_pos(current_element,     j)
	{
	for (j = current_element; j < current_no_of_spines; j++)
		{
		path_indicator[j] = path_indicator[j+1]
		current_interp[j] = current_interp[j+1]
		}
	delete path_indicator[j]
	delete current_interp[j]
	current_no_of_spines--
	} 

###################################################################
##				Function Exchange_spines
#	This function exhanges two spines by exchanging the corresponding
# elements in 'current_interp'
#
function exchange_spines(    j,arraya,count,temp)
	{
	count = 1
	for (j = 1; j <= current_no_of_spines; j += 1)
		{
		if (path_indicator[j] == "x")
			{
			arraya[count] = j
			path_indicator[j] = ""
			count += 1
			}
		}
	temp = current_interp[arraya[1]]
	current_interp[arraya[1]] = current_interp[arraya[2]]
	current_interp[arraya[2]] = temp
	}

#######################################################################
## 				Function Process_data
#	This function processes all tokens that belong to the target
# interpretation and makes use of the function recode to obtain the 
# appropriate resulting token.
#	
function process_data(     i,j,arrayb,split_num,current_token,current_line,
												return_token)
	{
	current_line = ""
	#
	# Loop through each spine
	#
	for (i = 1; i <= current_no_of_spines; i += 1)
		{
		current_token = ""
		#
		# If the current spine contains the target interpretation
		#
		if (current_interp[i] == target_interp)
			{
			split_num = split($i,arrayb," ")
			#
			# Loop through each part of a multiple stop
			#
			for (j = 1; j <= split_num; j += 1)
				{
				#
				# If the token contains the user defined regular expression
				# skip it
				#
				if (options ~ /s/ && arrayb[j] ~ reg_exp)
					current_token = current_token arrayb[j] " "
				#
				# Otherwise, process it
				#
				else
					{
					return_token = recode(arrayb[j])
					#
					# Make sure multiple stops don't have null tokens
					#
					if (return_token != "." || split_num == 1)
						current_token = current_token return_token " "
					}
				}
			#
			# Make sure at least a null token is output
			#
			if (current_token == "") current_line = current_line ".\t"
			else current_line = current_line \
					substr(current_token,1,length(current_token)-1) "\t"
			}
		#
		# If the current spine does not contain the target interpretation,
		# simply print it out
		#
		else
			{
			current_line = current_line $i "\t"
			}
		}
	#
	# Print the resulting data record
	#
	print substr(current_line,1,length(current_line)-1)
	}

##########################################################################
##				Function Recode
#	Called from process_data(), this function determines which condition
# the current token satisfies (from top of reassign file to bottom) and
# returns the appropriate token.
#
function recode(token,    i,return_token,value,matched_token)
	{
	return_token = ""
	value = ""
	#
	# If there is a number in the current token
	#
	if (match(token,number))
		{
		#
		# Get the number and loop through all of the conditions. As soon as
		# a condition is satisfied, the loop is broken
		#
		matched_token = substr(token,RSTART,RLENGTH)
		value = matched_token + 0
		sub(number,SUBSEP,token)
		for (i = 1; i <= num_of_conditions; i++)
			{
			if (relation[i] == "==")
				{
				if (value == condition[i])
					{
					return_token = action[i]
					break
					}
				}
			else if (relation[i] == "!=")
				{
				if (value != condition[i])
					{
					return_token = action[i]
					break
					}
				}
			else if (relation[i] == "<=")
				{
				if (value <= condition[i])
					{
					return_token = action[i]
					break
					}
				}
			else if (relation[i] == ">=")
				{
				if (value >= condition[i])
					{
					return_token = action[i]
					break
					}
				}
			else if (relation[i] == "<")
				{
				if (value < condition[i])
					{
					return_token = action[i]
					break
					}
				}
			else if (relation[i] == ">")
				{
				if (value > condition[i])
					{
					return_token = action[i]
					break
					}
				}
			}
		#
		# If there was no matching condition
		#
		if (return_token == "")
			{
			#
			# If there is a default output, return that
			#
			if (default_output != "")
				{
				if (options ~ /x/) return default_output
				else
					{
					sub(SUBSEP,default_output,token)
					return token
					}
				}
			#
			# Else, simply return the token
			#
			else
				{
				if (options ~ /x/) return matched_token
				else
					{
					sub(SUBSEP,matched_token,token)
					return token
					}
				}
			}
		#
		# Otherwise, return the resulting token
		#
		else
			{
			if (options ~ /x/) return return_token
			else
				{
				# This first gsub command is supposed to replace any
				# backslashes with 2 backslashes so that the sub command
				# that follows returns the correct number of backslashes
				# (It doesn't look like it should work but it does!)
				gsub(/\\/,"\\\\\\\\",return_token)
				sub(SUBSEP,return_token,token)
				return token
				}
			}
		}
	#
	# If there was no number in the token, return the entire token
	#
	else
		{
		if (options ~ /x/) return "."
		else return token
		}
	}
